modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpChannel.java

```diff
@@ -14,31 +14,26 @@
 import org.elasticsearch.common.util.concurrent.ListenableFuture;
 import org.elasticsearch.http.HttpChannel;
 import org.elasticsearch.http.HttpResponse;
-import org.elasticsearch.transport.TransportException;
-import org.elasticsearch.transport.netty4.Netty4TcpChannel;
 
 import java.net.InetSocketAddress;
 import java.net.SocketAddress;
 
+import static org.elasticsearch.transport.netty4.Netty4Utils.addListener;
+import static org.elasticsearch.transport.netty4.Netty4Utils.safeWriteAndFlush;
+
 public class Netty4HttpChannel implements HttpChannel {
 
     private final Channel channel;
     private final ListenableFuture<Void> closeContext = new ListenableFuture<>();
 
     Netty4HttpChannel(Channel channel) {
         this.channel = channel;
-        Netty4TcpChannel.addListener(this.channel.closeFuture(), closeContext);
+        addListener(this.channel.closeFuture(), closeContext);
     }
 
     @Override
     public void sendResponse(HttpResponse response, ActionListener<Void> listener) {
-        // We need to both guard against double resolving the listener and not resolving it in case of event loop shutdown so we need to
-        // use #notifyOnce here until https://github.com/netty/netty/issues/8007 is resolved.
-        var wrapped = ActionListener.notifyOnce(listener);
-        channel.writeAndFlush(response, Netty4TcpChannel.addPromise(wrapped, channel));
-        if (channel.eventLoop().isShutdown()) {
-            wrapped.onFailure(new TransportException("Cannot send HTTP response, event loop is shutting down."));
-        }
```

Comment:
Sorry for being slow on this issue. Could you please help me some more to understand what the issue was with the previous solution? You said

> We could still be manipulating a network message when the event loop shuts down, causing us to close the message while it's still in use.

What do you mean by "manipulating a network message"? Is it releasing the response (or "close the message")? I fail to see how we can "close the message while it's still in use" given the listener is notifyOnce?

Reply:
The problem is that `isShutdown()` returns true while the event loop is [still running its last few tasks](https://github.com/netty/netty/blob/467322e941b5437fbe2761e7ad41beead6ce28c3/common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java#L1029-L1041), and one of those tasks might still be doing something with the message while we release it. To be sure we're releasing the message after all possible uses on the event loop we need to subscribe to the `terminationFuture` instead.

Replacing the `releaseOnce` with `Promise#tryFailure` is not strictly necessary to fix this.

Reply:
Ah OK. That makes sense. I am curious how you found out about it. Did you just read into the Netty source code?

Reply:
Yes, your [review comment here](https://github.com/elastic/elasticsearch/pull/104851#discussion_r1487001145) prompted me to look a little deeper at the shutdown behaviour which made me realise we were doing it wrong in these places too.